unittests/forked_tests_if.cpp

```diff
@@ -0,0 +1,156 @@
+
+            // add this new block to our corrupted block merkle
+            fork.blocks.emplace_back(copy_b);
+         } else {
+            fork.blocks.emplace_back(b);
+         }
+      }
+   };
+
+   // First produce forks of 2 blocks on node3, so the fork switch will happen when we produce the
+   // third block which will have a newer timestamp than the last block of node0's branch.
+   // Finality progress is halted as the network is split, so the timestamp criteria decides the best fork.
+   //
+   // Skip producer prod with time delay (13 blocks)
+   // -----------------------------------------------------------------------------------------------------
+   for (size_t i = 0; i < num_forks-1; ++i) {
+      produce_and_store_block_on_node3_forks(i, 1);
+   }
+
+   // then produce 3 blocks on node0. This will be the default branch before we attempt to push the
+   // forks from node3.
+   // ---------------------------------------------------------------------------------------------
+   for (size_t i = 0; i < num_forks; ++i) {
+      auto sb = node0.produce_block(fc::milliseconds(config::block_interval_ms * (i==0 ? num_forks : 1)));
+      BOOST_REQUIRE_EQUAL(sb->producer, producers[prod]); // produced by the producer returned by `set_producers`
+   }
+
+   // Produce the last block of node3's forks, with a later timestamp than all 3 blocks of node0.
+   // When pushed to node0, It will cause a fork switch as it will be more recent than node0's head.
+   // -------------------------------------------------------
+   produce_and_store_block_on_node3_forks(num_forks-1, num_forks * 2);
+
+   // Now we push each fork (going from most corrupted fork to least) from node3 to node0.
+   // Blocks are correct enough to be pushed and inserted into fork_db, but will fail validation
+   // (when apply_block is called on the corrupted block). This will happen when the fork switch occurs,
+   // and all blocks from the forks are validated, which is why we expect an exception when the last
+   // block of the fork is pushed.
+   // -------------------------------------------------------------------------------------------------
+   for (size_t i = 0; i < forks.size(); i++) {
+      BOOST_TEST_CONTEXT("Testing Fork: " << i) {
+         const auto& fork = forks.at(i);
+         // push the fork to the original node
+         for (size_t fidx = 0; fidx < fork.blocks.size() - 1; fidx++) {
+            const auto& b = fork.blocks.at(fidx);
+            // push the block only if its not known already
+            if (!node0.control->fetch_block_by_id(b->calculate_id())) {
+               node0.push_block(b);
+            }
+         }
+
+         // push the block which should attempt the corrupted fork and fail
+         BOOST_REQUIRE_EXCEPTION( node0.push_block(fork.blocks.back()), fc::exception,
+                                  fc_exception_message_starts_with( "finality_mroot does not match" )
+         );
```

Comment:
Also, after each failure, check to see that the head block of node0 is the same as what it was before the test started pushing any of the bad forks.

Reply:
done.